pkg/controllers/infracluster/infracluster_controller.go

```diff
@@ -0,0 +1,288 @@
+				ControlPlaneEndpoint: clusterv1.APIEndpoint{
+					Host: apiUrl.Hostname(),
+					Port: int32(port),
+				},
+			},
+		}
+
+		if err := r.Create(ctx, &target); err != nil {
+			return fmt.Errorf("failed to create %q: %w", target.Kind, err)
+		}
+	} else if err != nil {
+		return fmt.Errorf("failed to get %q: %w", target.Kind, err)
+	}
+
+	r.InfraCluster = &target
+
+	return nil
+}
+
+// setAvailableCondition sets the ClusterOperator status condition to Available.
+func (r *InfraClusterController) setAvailableCondition(ctx context.Context, log logr.Logger) error {
+	co, err := r.GetOrCreateClusterOperator(ctx)
+	if err != nil {
+		return err
+	}
+
+	conds := []configv1.ClusterOperatorStatusCondition{
+		operatorstatus.NewClusterOperatorStatusCondition(InfraClusterControllerAvailableCondition, configv1.ConditionTrue, operatorstatus.ReasonAsExpected,
+			"InfraCluster Controller works as expected"),
+		operatorstatus.NewClusterOperatorStatusCondition(InfraClusterControllerDegradedCondition, configv1.ConditionFalse, operatorstatus.ReasonAsExpected,
+			"InfraCluster Controller works as expected"),
+	}
+
+	co.Status.Versions = []configv1.OperandVersion{{Name: controllers.OperatorVersionKey, Version: r.ReleaseVersion}}
+	log.V(2).Info("InfraCluster Controller is Available")
+	return r.SyncStatus(ctx, co, conds)
+}
+
+// setAvailableCondition sets the ClusterOperator status condition to Degraded.
+func (r *InfraClusterController) setDegradedCondition(ctx context.Context, log logr.Logger) error {
+	co, err := r.GetOrCreateClusterOperator(ctx)
+	if err != nil {
+		return err
+	}
+
+	conds := []configv1.ClusterOperatorStatusCondition{
+		operatorstatus.NewClusterOperatorStatusCondition(InfraClusterControllerAvailableCondition, configv1.ConditionFalse, operatorstatus.ReasonSyncFailed,
+			"InfraCluster Controller failed install"),
+		operatorstatus.NewClusterOperatorStatusCondition(InfraClusterControllerDegradedCondition, configv1.ConditionTrue, operatorstatus.ReasonSyncFailed,
+			"InfraCluster Controller failed install"),
+	}
+
+	co.Status.Versions = []configv1.OperandVersion{{Name: controllers.OperatorVersionKey, Version: r.ReleaseVersion}}
+	log.Info("InfraCluster Controller is Degraded")
+	return r.SyncStatus(ctx, co, conds)
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *InfraClusterController) SetupWithManager(mgr ctrl.Manager) error {
+	build := ctrl.NewControllerManagedBy(mgr).
+		For(&configv1.ClusterOperator{}, builder.WithPredicates(clusterOperatorPredicates()))
+		// TODO: write a function that based the Infra returns a watch to the cloud specific InfraClusters objects.
+
+	return build.Complete(r)
+}
+
+// setManagedByAnnotation is setting the ManagedBy Annotation.
+// When this annotation is set, Cluster API will ignore the InfraCluster object and assume it is managed by an external entity.
+func setManagedByAnnotation(annotations map[string]string) map[string]string {
```

Comment:
Maps are pointers so I don't think you need to return and reset this like this

Reply:
Yeah, we can apply annotations inline when we create the infraCluster directly.